src/vivarium_public_health/metrics/mortality.py

```diff
@@ -7,121 +7,173 @@
+        # todo can we specify only causes with excess mortality?
         diseases = builder.components.get_components_by_type(
             (DiseaseState, RiskAttributableDisease)
         )
-        self.causes = [c.state_id for c in diseases] + ["other_causes"]
-
-        life_expectancy_data = builder.data.load(
-            "population.theoretical_minimum_risk_life_expectancy"
-        )
-        self.life_expectancy = builder.lookup.build_table(
-            life_expectancy_data, key_columns=[], parameter_columns=["age"]
-        )
+        return [c.state_id for c in diseases] + ["other_causes"]
 
+    # noinspection PyMethodMayBeStatic
+    def _get_population_view(self, builder: Builder) -> PopulationView:
         columns_required = [
             "tracked",
             "alive",
-            "entrance_time",
-            "exit_time",
-            "cause_of_death",
             "years_of_life_lost",
-            "age",
+            "cause_of_death",
+            "exit_time",
         ]
-        if self.config.by_sex:
-            columns_required += ["sex"]
-        self.population_view = builder.population.get_view(columns_required)
-
-        builder.value.register_value_modifier("metrics", self.metrics)
+        return builder.population.get_view(columns_required)
 
-    def metrics(self, index, metrics):
-        pop = self.population_view.get(index)
-        pop.loc[pop.exit_time.isnull(), "exit_time"] = self.clock()
+    def _register_collect_metrics_listener(self, builder: Builder) -> None:
+        builder.event.register_listener("time_step", self.on_collect_metrics)
 
-        person_time = get_person_time(
-            pop, self.config.to_dict(), self.start_time, self.clock(), self.age_bins
-        )
-        deaths = get_deaths(
-            pop,
-            self.config.to_dict(),
-            self.start_time,
-            self.clock(),
-            self.age_bins,
-            self.causes,
-        )
-        ylls = get_years_of_life_lost(
-            pop,
-            self.config.to_dict(),
-            self.start_time,
-            self.clock(),
-            self.age_bins,
-            self.life_expectancy,
-            self.causes,
+    def _register_metrics_modifier(self, builder: Builder) -> None:
+        builder.value.register_value_modifier(
+            self.metrics_pipeline_name,
+            modifier=self.metrics,
+            requires_columns=["age", "exit_time", "alive"],
         )
 
-        metrics.update(person_time)
-        metrics.update(deaths)
-        metrics.update(ylls)
+    ########################
+    # Event-driven methods #
+    ########################
+
+    def on_collect_metrics(self, event: Event) -> None:
+        pop = self.population_view.get(event.index)
+        pop_died = pop[(pop["alive"] == "dead") & (pop["exit_time"] > event.time - self.step_size)]
+
+        groups = self.stratifier.group(
+            pop_died.index, set(self.config.include), set(self.config.exclude)
```

Comment:
Allow `group` to take lists as args and coerce to sets on the stratifier side.  Generally if you can make the code in the utility function a little uglier in one place to allow client code to be cleaner in a lot of places, you should do it.

Reply:
Idea for later (unless it's super easy and you agree for now): 

It may be nice to allow the user to provide an `as_mask` arg to `ResultsStratifier.group` that returns a boolean mask on the provided index to indicate the group.  This would support

```
groups = self.stratifier.group(
    pop_died.index, set(self.config.include), set(self.config.exclude)
)
for label, group_index in groups:
    for cause in self.causes_of_death:
        pop_died_of_cause = (
            pop_died.loc[group_index, :]
            .query(f"cause_of_death == '{cause}'")
        )
```
and 

```
groups = self.stratifier.group(
    pop_died.index, set(self.config.include), set(self.config.exclude), as_mask=True,
)
for label, group_mask in groups:
    for cause in self.causes_of_death:
        cause_mask = pop_died['cause_of_death'] == cause
        pop_died_of_cause = pop_died.loc[group_mask & cause_mask]
```
The mask syntax is going to allow for more flexibility (e.g. that doesn't rely on query syntax).  

It's also faster.  Simple example I have handy:

![image](https://user-images.githubusercontent.com/6313865/161176053-92f7cad8-7866-4e76-a6c4-cc941af57c99.png)

Reply:
Speedup is less profound if things are strings, so caveat: 
![image](https://user-images.githubusercontent.com/6313865/161176220-8563a3cb-6a4e-4c14-812f-1bd099ffc9ff.png)

Reply:
Works for me. I'm actually using masks in the other observers now that I look, so this will make things simpler there as well